client/rpc-servers/src/lib.rs

```diff
@@ -77,87 +80,49 @@ impl WsConfig {
-	metrics: Option<RpcMetrics>,
-	rpc_api: RpcModule<M>,
-	rt: tokio::runtime::Handle,
-) -> Result<HttpServerHandle, Box<dyn StdError + Send + Sync>> {
-	let max_payload_in = payload_size_or_default(max_payload_in_mb);
-	let max_payload_out = payload_size_or_default(max_payload_out_mb);
-
-	let mut acl = AccessControlBuilder::new();
-
-	if let Some(cors) = cors {
-		// Whitelist listening address.
-		// NOTE: set_allowed_hosts will whitelist both ports but only one will used.
-		acl = acl.set_allowed_hosts(format_allowed_hosts(&addrs[..]))?;
-		acl = acl.set_allowed_origins(cors)?;
-	};
-
-	let builder = HttpServerBuilder::new()
-		.max_request_body_size(max_payload_in as u32)
-		.max_response_body_size(max_payload_out as u32)
-		.set_access_control(acl.build())
-		.health_api("/health", "system_health")?
-		.custom_tokio_runtime(rt);
-
-	let rpc_api = build_rpc_api(rpc_api);
-	let (handle, addr) = if let Some(metrics) = metrics {
-		let middleware = RpcMiddleware::new(metrics, "http".into());
-		let builder = builder.set_middleware(middleware);
-		let server = builder.build(&addrs[..]).await?;
-		let addr = server.local_addr();
-		(server.start(rpc_api)?, addr)
-	} else {
-		let server = builder.build(&addrs[..]).await?;
-		let addr = server.local_addr();
-		(server.start(rpc_api)?, addr)
-	};
-
-	log::info!(
-		"Running JSON-RPC HTTP server: addr={}, allowed origins={:?}",
-		addr.map_or_else(|_| "unknown".to_string(), |a| a.to_string()),
-		cors
-	);
-
-	Ok(handle)
-}
-
-/// Start WS server listening on given address.
-pub async fn start_ws<M: Send + Sync + 'static>(
-	addrs: [SocketAddr; 2],
-	cors: Option<&Vec<String>>,
-	ws_config: WsConfig,
+	config: Config,
 	metrics: Option<RpcMetrics>,
 	rpc_api: RpcModule<M>,
 	rt: tokio::runtime::Handle,
 	id_provider: Option<Box<dyn IdProvider>>,
-) -> Result<WsServerHandle, Box<dyn StdError + Send + Sync>> {
+	transport_label: &str,
+) -> Result<ServerHandle, Box<dyn StdError + Send + Sync>> {
 	let (max_payload_in, max_payload_out, max_connections, max_subs_per_conn) =
-		ws_config.deconstruct();
-
-	let mut acl = AccessControlBuilder::new();
-
-	if let Some(cors) = cors {
-		// Whitelist listening address.
-		// NOTE: set_allowed_hosts will whitelist both ports but only one will used.
-		acl = acl.set_allowed_hosts(format_allowed_hosts(&addrs[..]))?;
-		acl = acl.set_allowed_origins(cors)?;
+		config.deconstruct();
+
+	let host_filter = hosts_filter(cors.is_some(), &addrs);
+
+	let cors = {
+		if let Some(cors) = cors {
+			let mut list = Vec::new();
+			for origin in cors {
+				list.push(HeaderValue::from_str(origin.as_str())?);
+			}
+			CorsLayer::new().allow_origin(AllowOrigin::list(list))
+		} else {
+			// allow all cors
+			CorsLayer::permissive()
+		}
 	};
 
-	let mut builder = WsServerBuilder::new()
+	let middleware = tower::ServiceBuilder::new()
+		// Proxy `GET /health` requests to internal `system_health` method.
+		.layer(ProxyGetRequestLayer::new("/health", "system_health")?)
```

Comment:
DQ: previously we exposed the `/health` API path for HTTP only. If we upgrade the HTTP connection to WS, would the `/health` API still be reachable for the "old-HTTP port, now WS port"?

Reply:
`/health` is enabled for both ports but as we discussed offline